File: src/types/product.ts
export type ProductBadge = "bestseller" | "new" | "limited" | "sale";

export type ProductSpec = {
  label: string;
  value: string;
};

export interface Product {
  id: string;
  slug: string;
  name: string;
  shortDescription: string;
  description: string;
  price: number;
  compareAtPrice?: number;
  currency: "USD";
  categorySlugs: string[];
  /**
   * Primary cover (kept for convenience).
   * Prefer `images` for carousels — always at least 1 item.
   */
  imageUrl: string;
  /** Gallery for product image carousel (resilient mock → real PDP later) */
  images: string[];
  imageAlt: string;
  rating: number;
  reviewCount: number;
  badges?: ProductBadge[];
  features: string[];
  /** PDP detail rows — Material / Design / Size style */
  specs?: ProductSpec[];
  inStock: boolean;
  featured?: boolean;
  categoryLabel?: string;
}

/** Specs for PDP table; falls back from features when missing */
export function getProductSpecs(product: Product): ProductSpec[] {
  if (product.specs && product.specs.length > 0) return product.specs;
  const labels = ["Design", "Build", "Care", "Detail"];
  return product.features.map((value, i) => ({
    label: labels[i] ?? `Detail ${i + 1}`,
    value,
  }));
}

export interface Category {
  id: string;
  slug: string;
  name: string;
  description: string;
  tagline: string;
  imageUrl: string;
  ctaLabel: string;
  productCount: number;
}

/** Normalize any product-like object to a non-empty image list. */
export function getProductImages(product: {
  imageUrl?: string;
  images?: string[];
}): string[] {
  const list = (product.images ?? []).filter(Boolean);
  if (list.length > 0) return list;
  if (product.imageUrl) return [product.imageUrl];
  return [];
}

File: backend/app/api/v1/schemas/product.py
"""Product / search schemas — mirrors src/types/product.ts."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class CamelModel(BaseModel):
    model_config = ConfigDict(
        populate_by_name=True,
        ser_json_by_alias=True,
    )


class ProductSpecOut(CamelModel):
    label: str
    value: str


class ProductOut(CamelModel):
    id: str
    slug: str
    name: str
    short_description: str = Field(serialization_alias="shortDescription")
    description: str
    price: float
    compare_at_price: float | None = Field(default=None, serialization_alias="compareAtPrice")
    currency: Literal["USD"] = "USD"
    category_slugs: list[str] = Field(serialization_alias="categorySlugs")
    image_url: str = Field(serialization_alias="imageUrl")
    images: list[str]
    image_alt: str = Field(serialization_alias="imageAlt")
    rating: float
    review_count: int = Field(serialization_alias="reviewCount")
    badges: list[str] | None = None
    features: list[str]
    specs: list[ProductSpecOut] | None = None
    in_stock: bool = Field(serialization_alias="inStock")
    stock_qty: int | None = Field(default=None, serialization_alias="stockQty")
    featured: bool | None = None
    category_label: str | None = Field(default=None, serialization_alias="categoryLabel")


class ProductDetailOut(CamelModel):
    """PDP payload: product + optional related list."""

    product: ProductOut
    related: list[ProductOut] = Field(default_factory=list)


class SearchResponseOut(CamelModel):
    query: str
    items: list[ProductOut]
    total: int
